File: lab/IS6404/4-ARIMA.py
import pandas as pd
import numpy as np
from statsmodels.tsa.arima.model import ARIMA
import matplotlib.pyplot as plt

# Đọc dữ liệu từ file CSV
data = pd.read_csv('./data/gdp.csv')

# Hiển thị dữ liệu để kiểm tra
print(data.head())
print(data.info())

# Chuyển đổi dữ liệu GDP thành chuỗi thời gian
# Định dạng index ngày tháng dựa trên thông tin năm
data['Year'] = pd.to_datetime(data['Year'], format='%Y')
data.set_index('Year', inplace=True)

# Phân tích chuỗi thời gian bằng mô hình ARIMA
# Lựa chọn tham số p, d, q cho mô hình ARIMA
model = ARIMA(data['GDP'], order=(1, 1, 1))  # Tham số order=(p, d, q) cần được lựa chọn dựa trên ACF và PACF
fit_model = model.fit()

# Dự báo cho 5 năm tiếp theo
forecast = fit_model.get_forecast(steps=5)
forecast_mean = forecast.predicted_mean
conf_int = forecast.conf_int()

# Hiển thị kết quả dự báo
print('Dự báo GDP:')
print(forecast_mean)

# Vẽ biểu đồ dự báo
plt.figure(figsize=(10, 5))
plt.plot(data.index, data['GDP'], label='Historical GDP')
plt.plot(pd.date_range(data.index[-1], periods=6, freq='A')[1:], forecast_mean, label='Forecasted GDP', color='red')
plt.fill_between(pd.date_range(data.index[-1], periods=6, freq='A')[1:], conf_int.iloc[:, 0], conf_int.iloc[:, 1], color='pink', alpha=0.3)
plt.title('GDP Forecast Using ARIMA')
plt.xlabel('Year')
plt.ylabel('GDP')
plt.legend()
plt.grid(True)
plt.show()

File: lab/IS6404 -T7/4-ARIMA.R
# Tải các thư viện cần thiết
library(forecast)

# Đọc dữ liệu từ file CSV
data <- read.csv("./data/gdp.csv")

# Hiển thị cấu trúc dữ liệu để xác nhận các cột và dữ liệu
print(head(data))
print(str(data))

# Chuyển đổi dữ liệu GDP thành chuỗi thời gian
gdp_ts <- ts(data$GDP, start = 1985, frequency = 1)  # Giả sử dữ liệu là hàng năm

# Phân tích chuỗi thời gian bằng mô hình ARIMA tự động
fit_arima <- auto.arima(gdp_ts)

# Dự báo cho 5 năm tiếp theo
forecast_arima <- forecast(fit_arima, h = 5)

# Hiển thị kết quả dự báo
print(forecast_arima)

# Vẽ biểu đồ dự báo
plot(forecast_arima)

# Thêm tiêu đề và nhãn cho biểu đồ
title(main = "Dự báo GDP bằng Mô Hình ARIMA", xlab = "Năm", ylab = "GDP")
legend("bottomleft", legend=c("Dự báo"), col=c("red"), lty=1, cex=0.8)
